mesh/mesh.go

```diff
@@ -550,6 +551,65 @@ func (msh *Mesh) revertState(ctx context.Context, logger log.Log, revertTo types
 	return nil
 }
 
+func (msh *Mesh) saveHareOutput(ctx context.Context, logger log.Log, lid types.LayerID, bid types.BlockID) error {
+	logger.Info("saving hare output for layer")
+	var (
+		certs []certificates.CertValidity
+		err   error
+	)
+	if err = msh.cdb.WithTx(ctx, func(tx *sql.Tx) error {
+		// check if a certificate has been sync'ed.
+		// hare outputs are processed in layer order. i.e. when hare fails for a previous layer N,
+		// a later layer N+x has to wait for syncer to fetch block certificate for the mesh to make
+		// progress. therefore, by the time layer N+x is processed, syncer may already have sync'ed
+		// the certificate for this layer. in this case, hare output no longer matters.
+		certs, err = certificates.Get(tx, lid)
+		if err != nil && !errors.Is(err, sql.ErrNotFound) {
+			return fmt.Errorf("mesh get cert %v: %w", lid, err)
+		}
+		saved := false
+		for _, c := range certs {
+			if c.Block == bid {
+				saved = true
+				break
+			} else if c.Valid {
+				err = certificates.SetHareOutputInvalid(tx, lid, bid)
+				if err != nil {
+					return fmt.Errorf("save invalid hare output %v/%v: %w", lid, bid, err)
+				}
+				saved = true
+				break
+			}
+		}
+		if !saved {
+			err = certificates.SetHareOutput(tx, lid, bid)
+			if err != nil {
+				return fmt.Errorf("save valid hare output %v/%v: %w", lid, bid, err)
+			}
+		}
+		return nil
+	}); err != nil {
+		return err
+	}
+	if len(certs) == 0 {
+		msh.trtl.OnHareOutput(lid, bid)
+	} else if len(certs) > 1 {
+		logger.With().Warning("multiple valid certs found in network",
+			log.Object("certificates", log.ObjectMarshallerFunc(func(encoder zapcore.ObjectEncoder) error {
+				for _, cert := range certs {
+					encoder.AddString("block_id", cert.Block.String())
+					encoder.AddBool("valid", cert.Valid)
+				}
+				return nil
+			})))
+	} else {
+		logger.With().Info("already synced certificate",
+			log.Stringer("cert_block_id", certs[0].Block),
+			log.Bool("cert_valid", certs[0].Valid))
+	}
```

Comment:
NIT: i think switch would help with readability here

```suggestion
	switch len(certs) {
	case 0:
		msh.trtl.OnHareOutput(lid, bid)
	case 1:
		logger.With().Info("already synced certificate",
			log.Stringer("cert_block_id", certs[0].Block),
			log.Bool("cert_valid", certs[0].Valid))
	default: // more than 1 cert
		logger.With().Warning("multiple valid certs found in network",
			log.Object("certificates", log.ObjectMarshallerFunc(func(encoder zapcore.ObjectEncoder) error {
				for _, cert := range certs {
					encoder.AddString("block_id", cert.Block.String())
					encoder.AddBool("valid", cert.Valid)
				}
				return nil
			})))
	}
```